measurements/pages/__init__.py

```diff
@@ -204,34 +208,84 @@ def files_in_country(country_code):
+                    if (streamed_data + len(chunk)) > report_size:
+                        chunk = chunk[:abs(streamed_data - report_size)]
+                    yield chunk
+                    streamed_data += len(chunk)
+        except Exception as exc:
+            raise HTTPException("Failed to fetch data: %s" % exc)
+    return generator
+
+@pages_blueprint.route('/files/download/<path:textname>')
+def files_download(textname):
+    subquery = current_app.db_session.query(
+            Measurement.frame_off.label('frame_off'),
+            Measurement.frame_size.label('frame_size'),
+            Measurement.intra_off.label('intra_off'),
+            Measurement.intra_size.label('intra_size'),
+            func.row_number().over(order_by='frame_off, intra_off').label('row_number'),
+            func.count().over().label('total_count'),
+            func.sum(Measurement.intra_size).over().label('report_size'),
+            Autoclaved.filename.label('filename'),
+    ).filter(Report.textname == textname) \
+        .join(Report, Report.report_no == Measurement.report_no) \
+        .join(Autoclaved, Autoclaved.autoclaved_no == Report.autoclaved_no) \
+        .subquery()
 
-@pages_blueprint.route('/files/download/<filename>')
-def files_download(filename):
-    try:
-        report = current_app.db_session.query(Report) \
-                        .filter(Report.textname.like("%" + filename)).first()
-        # XXX
-        # We have duplicate measurements :(
-        # So the below exception actually happens. This should be resolved
-        # in the data processing pipeline.
-        #   Report.filename == filename).one()
-    except NoResultFound:
-        raise NotFound("No file with that filename found")
-    except MultipleResultsFound:
-        # This should never happen.
-        raise HTTPException("Duplicate measurement detected")
-
-    if report is None:
+    q = current_app.db_session.query(
+        subquery.c.frame_off,
+        subquery.c.frame_size,
+        subquery.c.intra_off,
+        subquery.c.intra_size,
+        subquery.c.row_number.label('c_row_number'),
+        subquery.c.total_count.label('c_total_count'),
+        subquery.c.report_size,
+        subquery.c.filename
+    ).filter(or_(subquery.c.row_number == 1, subquery.c.row_number == subquery.c.total_count))
+
+    msmts = q.all()
+    if len(msmts) == 0:
+        current_app.logger.debug("Could not find %s" % textname)
         raise NotFound("No file with that filename found")
 
-    bucket_date, filename = report.textname.split("/")
-    return redirect(
-        get_download_url(
-            current_app,
-            bucket_date,
-            filename
-        )
-    )
+    autoclaved_filename = msmts[0].filename
+    intra_off = msmts[0].intra_off
+    frame_off = msmts[0].frame_off
+    total_frame_size = msmts[-1].frame_off - msmts[0].frame_off + msmts[-1].frame_size
+    report_size = msmts[0].report_size + 1 # We add one for the final newline
```

Comment:
`sum(intra_size + 1)` != `sum(intra_size) + 1`

Reply:
Ah right, because the `intra_size` is not inclusive of the `\n` character. Will adjust accordingly.